Blob range request test failures

annevk opened the pull request:
#### 09841b4faec73a9d1b15b37d68941a995f33cadf
<pre>
Blob range request test failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=248994">https://bugs.webkit.org/show_bug.cgi?id=248994</a>
rdar://103171187

Reviewed by Youenn Fablet.

This changes the Range request header parser so it can be used by both blob: URLs and the CORS-safelisted request-header check.

It changes the error handling for blob: URLs such that it now returns network errors rather than HTTP responses.

It aligns the Range request header parser with the Fetch Standard.

* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any.js:
(async string_appeared_here):
* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any.js: Added.
(supportedBlobRange.forEach):
(async unsupportedBlobRange):
(async unsupportedBlobRange.forEach):

These tests are from upstream web-platform-tests. <a href="https://github.com/web-platform-tests/wpt/pull/39108">https://github.com/web-platform-tests/wpt/pull/39108</a> added coverage for some of the cases discussed during review.

* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandle::doStart):
(WebCore::BlobResourceHandle::seek):
(WebCore::BlobResourceHandle::notifyResponse):
(WebCore::BlobResourceHandle::notifyResponseOnSuccess):
(WebCore::BlobResourceHandle::notifyResponseOnError): Deleted.
* Source/WebCore/platform/network/BlobResourceHandle.h:
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::isValidAcceptHeaderValue):
(WebCore::filenameFromHTTPContentDisposition):
(WebCore::isTabOrSpace):
(WebCore::parseRange):
(WebCore::isCrossOriginSafeRequestHeader):
(WebCore::isSimpleRangeHeaderValue): Deleted.
* Source/WebCore/platform/network/HTTPParsers.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::seek):
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:

Canonical link: <a href="https://commits.webkit.org/261968@main">https://commits.webkit.org/261968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9a11696bc0220c59124837891bb6dbbd37de4f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/138 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/121 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/132 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/116 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/131 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/125 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/112 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/128 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/129 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
```diff
@@ -223,89 +218,66 @@ void NetworkDataTaskBlob::seek()
+void NetworkDataTaskBlob::dispatchDidReceiveResponse()
 {
-    LOG(NetworkSession, "%p - NetworkDataTaskBlob::dispatchDidReceiveResponse(%u)", this, static_cast<unsigned>(errorCode));
+    LOG(NetworkSession, "%p - NetworkDataTaskBlob::dispatchDidReceiveResponse(%u)", this, static_cast<unsigned>(Error::NoError));
```
Is this still useful now this is only invoked upon success?

youennf replied:
Nope, we should probably update it

annevk commented on Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
```diff
@@ -223,89 +218,66 @@ void NetworkDataTaskBlob::seek()
+    // Notably, this will affect a name suggested in "File Save As".
 
-    didReceiveResponse(WTFMove(response), NegotiatedLegacyTLS::No, PrivateRelayed::No, [this, protectedThis = WTFMove(protectedThis), errorCode](PolicyAction policyAction) {
+    didReceiveResponse(WTFMove(response), NegotiatedLegacyTLS::No, PrivateRelayed::No, [this, protectedThis = WTFMove(protectedThis)](PolicyAction policyAction) {
```
Was this adjusted correctly? I initially substituted `errorCode` with `Error::NoError`, but that did not work.

youennf replied:
This is correct, a lambda is storing a variable with the provided name. Error::NoError is a value without a valid name

youennf commented on Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
```diff
@@ -223,89 +218,66 @@ void NetworkDataTaskBlob::seek()
+    response.setHTTPHeaderField(HTTPHeaderName::ContentLength, String::number(m_totalRemainingSize));
+    addPolicyContainerHeaders(response, m_blobData->policyContainer());
+
+    if (m_isRangeRequest) {
```
For oneliners, we do not use `{`and `}`

youennf commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -211,9 +205,9 @@ void BlobResourceHandle::doStart()
-    if (!range.isEmpty() && !parseRange(range, m_rangeOffset, m_rangeEnd, m_rangeSuffixLength)) {
-        m_errorCode = Error::RangeError;
-        notifyResponse();
+    m_isRangeRequest = !range.isNull();
```
This change (a request with an empty Range header) is probably worth a test.

youennf commented on Source/WebCore/platform/network/BlobResourceHandle.cpp:
```diff
@@ -558,9 +556,7 @@ void BlobResourceHandle::notifyResponse()
-        notifyResponseOnError();
-        notifyFinish();
+        notifyFail(m_errorCode);
     } else
```
Either remove the `{`and `}` or you can write it as:
```
if (foo) {
    notify...();
    return;
}
notify...();```

youennf commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -610,32 +610,49 @@ OptionSet<ClearSiteDataValue> parseClearSiteDataHeader(const ResourceResponse& r
+    // FIXME: this whole function could be more efficient by walking through the range value once.
+    if (!allowWhitespace) {
+        // Only 0x20 and 0x09 matter as newlines are already gone by the time we parse a header value.
+        index = range.find(' ');
```
It is a bit confusing to use the function scoped index variable.
I would just write it: `if (range.find(' ') != -1 return false;`
There is a find which takes an invocable, you might be able to use it to remove the FIXME.

youennf commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -976,11 +958,14 @@ bool isCrossOriginSafeRequestHeader(HTTPHeaderName name, const String& value)
-        if (!isSimpleRangeHeaderValue(value))
+        long long start;
+        long long end;
+        if (!parseRange(value, false, start, end))
```
It is not easy to understand what false means here.
We tend to prefer boolean enums so that it would read like:
`parseRange(value, AllowWhitespace::Yes, start, end)`

youennf reviewed (commented):
It would be good to add a test for the null Range header value and have the commit message with the right formatting and explaining with some more details what this patch is doing.
Some nits below.

youennf commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -610,20 +610,31 @@ OptionSet<ClearSiteDataValue> parseClearSiteDataHeader(const ResourceResponse& r
+    if (!startsWithLettersIgnoringASCIICase(range, "bytes"_s))
+        return false;
+
+    StringView byteRange = stripLeadingAndTrailingHTTPSpaces(range.substring(bytesLength));
```
auto?

youennf commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -610,20 +610,31 @@ OptionSet<ClearSiteDataValue> parseClearSiteDataHeader(const ResourceResponse& r
+
+    StringView byteRange = stripLeadingAndTrailingHTTPSpaces(range.substring(bytesLength));
+
+    if (byteRange.characterAt(0) != '=')
```
What is the guarantee byteRange is not empty?
Do we have a test with `Range: bytes  `?

annevk replied:
Ouch, this was a crash. Switched to `startsWith()` and added coverage.

youennf commented on Source/WebCore/platform/network/HTTPParsers.cpp:
```diff
@@ -610,20 +610,31 @@ OptionSet<ClearSiteDataValue> parseClearSiteDataHeader(const ResourceResponse& r
+        if (range.find(' ') != notFound)
+            return false;
+        if (range.find('\t') != notFound)
+            return false;
```
Would the following work:
```
static bool isTabOrSpace(UChar value)
{
    return value == '\t'|| value == ' ';
}

bool parseRange(...)
{
    ....
    if (allowWhitespace == RangeAllowWhitespace::No && range.find(isTabOrSpace) != notFound)
        return false;
}
```

annevk replied:
I will try that, but what I meant with my FIXME was turning `parseRange` more into something like https://fetch.spec.whatwg.org/#simple-range-header-value where you only iterate through the whole range once. That would be a lot easier if the `skipWhile` methods took a `StringView`.

marcoscaceres-remote commented on LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any.js:
```diff
@@ -0,0 +1,208 @@
+      return resp.text();
+    }).then(function(text) {
+      assert_equals(text, result, "Response's body is correct");
+    });
```
This is labelled `async` but it's not using `await`? 

Could we change it to the following and then send it back upstream? (alternatively, we can fix it in WPT later if we can be bothered)

```suggestion
    const blob = new Blob(data, { "type" : type });
    const blobURL = URL.createObjectURL(blob);
    t.add_cleanup(() => URL.revokeObjectURL(blobURL));
    const resp = await fetch(blobURL, {
      "headers": {
        "Range": range
      }
    });
    assert_equals(resp.status, 206, "HTTP status is 206");
    assert_equals(resp.type, "basic", "response type is basic");
    assert_equals(resp.headers.get("Content-Type"), type, "Content-Type is " + resp.headers.get("Content-Type"));
    assert_equals(resp.headers.get("Content-Length"), content_length.toString(), "Content-Length is " + resp.headers.get("Content-Length"));
    assert_equals(resp.headers.get("Content-Range"), content_range, "Content-Range is " + resp.headers.get("Content-Range"));
    const text = await resp.text();
    assert_equals(text, result, "Response's body is correct");
```

annevk replied:
Fair updated here and in https://github.com/web-platform-tests/wpt/pull/39108.

youennf reviewed (approved):
It would be good to update the commit message to state that these added/updated tests are just resynced from WPT upstream.

annevk committed:
Blob range request test failures
https://bugs.webkit.org/show_bug.cgi?id=248994
rdar://103171187

Reviewed by Youenn Fablet.

This changes the Range request header parser so it can be used by both blob: URLs and the CORS-safelisted request-header check.

It changes the error handling for blob: URLs such that it now returns network errors rather than HTTP responses.

It aligns the Range request header parser with the Fetch Standard.

* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any.js:
(async string_appeared_here):
* LayoutTests/imported/w3c/web-platform-tests/cors/cors-safelisted-request-header.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fetch/range/blob.any.js: Added.
(supportedBlobRange.forEach):
(async unsupportedBlobRange):
(async unsupportedBlobRange.forEach):

These tests are from upstream web-platform-tests. https://github.com/web-platform-tests/wpt/pull/39108 added coverage for some of the cases discussed during review.

* Source/WebCore/platform/network/BlobResourceHandle.cpp:
(WebCore::BlobResourceHandle::doStart):
(WebCore::BlobResourceHandle::seek):
(WebCore::BlobResourceHandle::notifyResponse):
(WebCore::BlobResourceHandle::notifyResponseOnSuccess):
(WebCore::BlobResourceHandle::notifyResponseOnError): Deleted.
* Source/WebCore/platform/network/BlobResourceHandle.h:
* Source/WebCore/platform/network/HTTPParsers.cpp:
(WebCore::isValidAcceptHeaderValue):
(WebCore::filenameFromHTTPContentDisposition):
(WebCore::isTabOrSpace):
(WebCore::parseRange):
(WebCore::isCrossOriginSafeRequestHeader):
(WebCore::isSimpleRangeHeaderValue): Deleted.
* Source/WebCore/platform/network/HTTPParsers.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::resume):
(WebKit::NetworkDataTaskBlob::seek):
(WebKit::NetworkDataTaskBlob::dispatchDidReceiveResponse):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:

Canonical link: https://commits.webkit.org/261968@main
